massa-execution-worker/src/execution.rs

```diff
@@ -248,7 +242,75 @@ impl ExecutionState {
         if slot > max_slot {
             panic!("cannot execute at a slot beyond active cursor + 1");
         }
+    }
+
+    /// Lazily query (from end to beginning) the active balance of an address at a given slot.
+    /// Returns None if the address balance could not be determined from the active history.
+    pub fn get_active_balance_at_slot(&self, slot: Slot, addr: &Address) -> Option<Amount> {
+        self.verify_active_slot(slot);
+
+        let iter = self
+            .active_history
+            .iter()
+            .rev()
+            .skip_while(|output| output.slot >= slot);
```

Comment:
this has linear complexity

note that it can be implemented with constant complexity because active_history elements represent strictly consecutive slots in increasing order